src/main/java/org/myrobotlab/service/abstracts/AbstractMotorController.java

```
@@ -1,17 +1,224 @@
+    double minY = (mc.getMinY() == null)? this.minY:mc.getMinY();
+    double maxY = (mc.getMaxY() == null)? this.maxY:mc.getMaxY();
+    
+    // motor control has highest precedence
+    Double minInput = mc.getMinInput();
+    Double maxInput = mc.getMaxInput();
+    // motor controller has next order of precedence
+    if (minInput == null) {
+      minInput = this.minInput;
+    }
+    if (maxInput == null) {
+      maxInput = this.maxInput;
+    }
+
+    // motor control has highest precedence
+    Double minOutput = mc.getMinOutput();
+    Double maxOutput = mc.getMaxOutput();
+    // motor controller has next order of precedence
+    if (minOutput == null) {
+      minOutput = this.minOutput;
+    }
+    if (maxOutput == null) {
+      maxOutput = this.maxOutput;
+    }
+    
+    if (minInput != null && in < minInput) {
+      log.warn("clipping input {} to {}", in, minInput);
+      in = minInput;
+    }
+    if (maxInput != null && in > maxInput) {
+      log.warn("clipping input {} to {}", in, maxInput);
+      in = maxInput;
+    }
+    
+    double c = minY + ((in - minX) * (maxY - minY)) / (maxX - minX);
+
+    if (minOutput != null && c < minOutput) {
+      log.warn("clipping output {} to {}", c, minOutput);
+      return minOutput;
+    }
+    if (maxOutput != null && c > maxOutput) {
+      log.warn("clipping output {} to {}", c, maxOutput);
+      return maxOutput;
+    }
+    return c;
+  }
+  
+  // FIXME - if kept they should be put in interface
+  public boolean motorDetach(String name) {
+    if (motors.containsKey(name)) {
+      motors.remove(name);
+      return true;
+    }
+    return false;
+  }
+
+  // FIXME - if kept they should be put in interface
+  public void motorMove(String name) {
+    motorMove((MotorControl) Runtime.getService(name));
+  }
+
+  // FIXME - if kept they should be put in interface
+  public void motorMoveTo(String name, Integer position) {
+    error("not implemented");
+  }
+  
+  @Override
```

should throw an unimplemented exception or they should have an implementation

---

maybe ... this shows a nice red glowing error bar in the gui, which might be nicer to teaching a new user that continuous dc motors without feedback cannot be movedTo